### PR TITLE
support specify --type for create-remote-secret command

### PR DESCRIPTION
**Pull request by irisdingbj**
For external Istiod scenario, we will use `istio-kubeconfig` as secret but now `create-remote-secert` will use the fixed prefix `istio-remote-secret-` for the create secret.

Add this flag so this command can be used to support external Istiod as well.

### CODE REVIEW

**Review by ericvn — APPROVED** [review]

**Comment by nmittler** [comment]
@irisdingbj I'm not exactly sure why we need a new naming convention exactly.

For remote clusters, we use `istio-remote-secret-<cluster>`. This secret already contains an annotation indicating the cluster name, so the fact that we're including the cluster in the name is really just to provide a unique name.  Since we'll likely use a remote cluster as the source of configuration for an external control plane, we don't we just add an additional annotation to the secret that indicates it's a configuration cluster?

@stevenctl WDYT?

**Comment by stevenctl** [comment]
I think the what was discussed in the 2 docs about labeling/annotating secrets to identify them as the config cluster would be better than requiring a special name. 

This comes along with some refactoring to Pilot bootstrap code to re-use the SecretController logic for the config cluster, so it's a bit of work. In the meantime for testing this command with external control-plane, simply renaming the secret by hand is better than adding this in a user facing place.

**Comment by irisdingbj** [comment]
@nmittler @stevenctl I need give more background here.  
Istiod on external control plane need to use `istio-kubeconfig` secret  to point its KUBECONFIG to the config cluster.  In the past we are using KUBECONFIG for kubectl directly and it might be short lived. Now we can enhance the `create-remote-secret` command to generate `istio-kubeconfig` secret for a specified sa to fullfill this. 

So I add the  `--secret-name` to define the generate secret name instead of the fixed `istio-remote-secret-'cluster name`"

**Comment by stevenctl** [comment]
Users wouldn't edit. We would if we want to experiment. Instead, we need to refactor things so we don't bother with the special mounted secret.

**Comment by irisdingbj** [comment]
> Users wouldn't edit. We would if we want to experiment. Instead, we need to refactor things so we don't bother with the special mounted secret. Rather than a specific secret name, can we watch all files in a directory and mount to that directory, for the initial secret?

I am working on a following up PR which will use the `istio-kubeconfig` secret generated by this command.  We will also watch this secret.  In case the token in the secret get updated, we will use the updated token automatically in our client.   User can choose to use current mounted secret if they like ( for backward compatibility).

**Comment by irisdingbj** [comment]
@stevenctl @nmittler  this is the PR which will use the secret generated here. https://github.com/istio/istio/pull/27465

File: istioctl/pkg/multicluster/remote_secret.go [hunk 1]
```diff
@@ -338,6 +344,9 @@ func (o *RemoteSecretOptions) addFlags(flagset *pflag.FlagSet) {
 	flagset.StringToString("auth-plugin-config", o.AuthPluginConfig,
 		fmt.Sprintf("Authenticator plug-in configuration. --auth-type=%v must be set with this option",
 			RemoteSecretAuthTypePlugin))
+	flagset.StringVar(&o.SecretName, "secret-name", "",
+		"Create a secret with this name. If secret-name is not specified, "+
+			remoteSecretPrefix+"\"name of the cluster\" will be used.")
```
Comment (esnible) [inline istioctl/pkg/multicluster/remote_secret.go:349]: Perhaps `Secret Name. (default: " + remoteSecretPrefix + "<name>)"`?

**Comment by nmittler** [comment]
@irisdingbj I agree with @stevenctl here... I'm not sure this is the right approach. I understand that you're just building on what already exists, but the external control plane logic is still WIP and will not be supported for 1.8 anyway. I think now is the right time to discuss what direction we want to go with all of this.

Perhaps we should have a VC to discuss?

FYI @linsun

**Comment by linsun** [comment]
@stevenctl @nmittler The motivation of this PR was brought up in the env WG last week via [this doc](https://docs.google.com/document/d/1Mt-RzOf2nny2D1j8BFM78M2ZvRLwCISWzbRWZvx-6qE/edit)
The consensus from the WG is we want to extend create-remote-secret to allow mesh operators to easily create secret for config cluster.  

There is only 1 config cluster per external Istiod, and that secret name is not configurable today, which is `istio-kubeconfig`.  I discussed with @irisdingbj offline to not expose `--secret-name` but instead allow mesh operators to specify if they are creating remote secrets for config cluster (or remote cluster which is default) via cmd like:

```
istioctl x create-remote-secret --type=config --service-account istiod-service-account
```
There is no need to specify secret name in this case, which should help addressing your concerns earlier.  I hope this clarifies things.  Feel free to let us know if a VC is still needed.

**Comment by irisdingbj** [comment]
@esnible Thanks for the review, PTAL as I have change to `--type` instead of `--secret-name`
@nmittler @stevenctl @linsun PTAL

**Comment by stevenctl** [comment]
We shouldn't couple the cluster name and whether or not its a config cluster. There are other changes that can be done along with this that enable setting the cluster name for the config cluster without relying on `values.global` generating flags/env vars to istiod. 

The only thing the type flag should do is add the label to the generated secret. We should default to `Service` type clusters if the `--type` flag isn't specified in istioctl, or if the label is missing. and spin up k8s service registries for them.

**Comment by linsun** [comment]
Agreed @stevenctl , cluster name isn't configurable when it is config cluster.  longer term, when we support multiple config clusters, we could make the cluster name configurable.  At the moment, it is not supported.

**Comment by stevenctl** [comment]
Just so I understand:
- If we have `--type` has Config, we give the secret a special name
- That special name causes us to use the secret the way we do today

We should solidify the design in the doc, I don't feel like I see the long term picture for how this will end up.

**Comment by linsun** [comment]
your understanding  is correct @stevenctl.

Agree with solidify this in design doc for longer term plan.

**Comment by irisdingbj** [comment]
@esnible PTAL, Thanks!

**Review by esnible — APPROVED** [review]

**Pull request merged**